### PR TITLE
[STORM-3924]  Support for declaring WorkerHook in Flux topology definitions

### PR DESCRIPTION
**Pull request by 6harat**
## What is the purpose of the change

A while back the support for declaring WorkerHook in TopologyBuilder was added. Related JIRA: https://issues.apache.org/jira/browse/STORM-126 (related PR: https://github.com/apache/storm/pull/884). This change adds support for the aforementioned feature for topology definitions submitted as Flux yaml.

## How was the change tested

Correspoding feature test case was added in `org.apache.storm.flux.TCKTest`
For regression: Git Actions build and test run as part of this pull request

### CODE REVIEW

**Comment by 6harat** [comment]
Please note the following observations / callouts:

1. have not added workerHoosk to executioncontext. AFAIS constructed spouts and bolts are being stored here so as to allow creation of streams later in the flow `FluxBuilder.buildStreamDefinitions`
2. have not added explicit type descriptors as being done for spouts, bolts and includes in `FluxParser.yaml`. AFAIU the info provided in TopologyDef is more than enough to parse yaml containing `workerHooks` field.

Let me know if my understanding was incorrect and a change is indeed required on the above files. Will pick it up right away.

**Comment by 6harat** [comment]
I was going through other parts of the code and found this TODO. Seems like it was intended to be connected to WorkerHooks when those were first introduced.

```
    private Map<String, Object> makeUserResources() {
        /* TODO: need to invoke a hook provided by the topology, giving it a chance to create user resources.
         * this would be part of the initialization hook
         * need to separate workertopologycontext into WorkerContext and WorkerUserContext.
         * actually just do it via interfaces. just need to make sure to hide setResource from tasks
         */
        return new HashMap<>();
    }
```

AFAIS, the current implementation will always lead to empty userResources as no interface exposes a way to allow user to set them.

Lmk, if my above understanding is correct and this change can also be taken up. this would allow for better application context management. Will put this fix in a separate PR.

cc: @abhishekagarwal87 , @schonfeld , @revans2 , @nathanmarz , @rzo1 , @bipinprasad

**Comment by 6harat** [comment]
> I was going through other parts of the code and found this TODO. Seems like it was intended to be connected to WorkerHooks when those were first introduced.
> 
> ```
>     private Map<String, Object> makeUserResources() {
>         /* TODO: need to invoke a hook provided by the topology, giving it a chance to create user resources.
>          * this would be part of the initialization hook
>          * need to separate workertopologycontext into WorkerContext and WorkerUserContext.
>          * actually just do it via interfaces. just need to make sure to hide setResource from tasks
>          */
>         return new HashMap<>();
>     }
> ```
> 
> AFAIS, the current implementation will always lead to empty userResources as no interface exposes a way to allow user to set them.
> 
> Lmk, if my above understanding is correct and this change can also be taken up. this would allow for better application context management. Will put this fix in a separate PR.
> 
> cc: @abhishekagarwal87 , @schonfeld , @revans2 , @nathanmarz , @rzo1 , @bipinprasad

this has been taken up here: https://github.com/apache/storm/pull/3547

File: docs/flux.md [hunk 1]
```diff
@@ -765,6 +770,16 @@ The `override` property controls how includes affect the values defined in the c
 
 **N.B.:** Includes are not yet recursive. Includes from included files will be ignored.
 
+## Worker Hooks
+Flux allows you to attach topology component that can be executed when a worker starts, and when a worker shuts down. It can be useful when you want to execute operations before topology processing starts, or cleanup operations before your workers shut down, e.g. managing application context. Worker Hooks should be an implementation of [IWorkerHook]({{page.git-blob-base}}/storm-client/src/jvm/org/apache/storm/hooks/IWorkerHook.java). Other than that, they follow similar Bean definition semantics as [Components](##Components) for declaration within yaml file.
```
Comment (rzo1) [inline docs/flux.md:774]: `components`

**Review by rzo1 — APPROVED** [review]

File: docs/flux.md (hunk 1 above)
Comment (6harat) [reply docs/flux.md]: rectified the typo in the docs

**Comment by 6harat** [comment]
@bipinprasad just pushed a README update commit, the prev workflow was already successful, so triggering it again might not be reqd. Please let me know if anything else is required from my side for going ahead with the merge.

**Comment by rzo1** [comment]
@6harat triggered the workflow.

**Comment by rzo1** [comment]
@6harat Can you try a rebase? So we can find out why the license file checks fail on your branch?

**Comment by 6harat** [comment]
will rebase and check it out over the weekend

**Comment by rzo1** [comment]
Any chance you can rebase @6harat ?

**Comment by rzo1** [comment]
I did the rebase / merge, so we get new CI results. Your changes shouldn't affect the license files, so we are good to move forwards soon. Thanks for this contribution!

**Comment by 6harat** [comment]
@rzo1 thanks a lot for rebasing it. Had been busy at ofc for the past couple of weeks, so couldn't do it earlier as planned. Let me know if anything else is required here.

**Pull request merged**